pkg/graph/constructor.go

```diff
@@ -17,14 +17,155 @@ limitations under the License.
+			logger.Warn("failed to list brokers while constructing lineage graph", zap.Error(err))
+		}
+
+		if err == nil {
+			for _, broker := range brokers.Items {
+				if config.ShouldAddBroker(broker) {
+					g.AddBroker(broker)
+				}
+			}
+		}
+
+		channels, err := eventingClient.MessagingV1().Channels(ns).List(ctx, metav1.ListOptions{})
+		if err != nil && !apierrs.IsNotFound(err) && !apierrs.IsUnauthorized(err) && !apierrs.IsForbidden(err) {
+			return nil, err
+		}
+
+		if apierrs.IsUnauthorized(err) || apierrs.IsForbidden(err) {
+			logger.Warn("failed to list channels while constructing lineage graph", zap.Error(err))
+		}
+
+		if err == nil {
+			for _, channel := range channels.Items {
+				if config.ShouldAddChannel(channel) {
+					g.AddChannel(channel)
+				}
+			}
+		}
+
+		sources, err := getSources(ctx, config)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, source := range sources {
+			if config.ShouldAddSource(source) {
+				g.AddSource(source)
+			}
+		}
+
+		triggers, err := eventingClient.EventingV1().Triggers(ns).List(ctx, metav1.ListOptions{})
+		if err != nil && !apierrs.IsNotFound(err) && !apierrs.IsUnauthorized(err) && !apierrs.IsForbidden(err) {
+			return nil, err
+		}
+
+		if apierrs.IsUnauthorized(err) || apierrs.IsForbidden(err) {
+			logger.Warn("failed to list triggers while constructing lineage graph", zap.Error(err))
+		}
+
+		if err == nil {
+			for _, trigger := range triggers.Items {
+				if config.ShouldAddTrigger(trigger) {
+					err := g.AddTrigger(trigger)
+					if err != nil {
+						return nil, err
+					}
+				}
+			}
+		}
+
+		subscriptions, err := eventingClient.MessagingV1().Subscriptions(ns).List(ctx, metav1.ListOptions{})
+		if err != nil && !apierrs.IsNotFound(err) && !apierrs.IsUnauthorized(err) && !apierrs.IsForbidden(err) {
+			return nil, err
+		}
+
+		if apierrs.IsUnauthorized(err) || apierrs.IsForbidden(err) {
+			logger.Warn("failed to list subscriptions while constructing lineage graph", zap.Error(err))
+		}
+
+		if err == nil {
+			for _, subscription := range subscriptions.Items {
+				if config.ShouldAddSubscription(subscription) {
+					err := g.AddSubscription(subscription)
+					if err != nil {
+						return nil, err
+					}
+				}
+			}
+		}
+
+		eventTypes, err := eventingClient.EventingV1beta3().EventTypes(ns).List(ctx, metav1.ListOptions{})
+		if err != nil && !apierrs.IsNotFound(err) && !apierrs.IsUnauthorized(err) && !apierrs.IsForbidden(err) {
+			return nil, err
+		}
+
+		if apierrs.IsUnauthorized(err) || apierrs.IsForbidden(err) {
+			logger.Warn("failed to list eventtypes while constructing lineage graph", zap.Error(err))
+		}
+
+		if err == nil {
+			for _, eventType := range eventTypes.Items {
+				if config.ShouldAddEventType(eventType) {
+					err := g.AddEventType(eventType)
+					if err != nil {
+						return nil, err
+					}
+				}
+			}
+		}
+	}
+
+	return g, nil
+}
+
 func (g *Graph) AddBroker(broker eventingv1.Broker) {
```

Comment:
Can we actually embed the full broker resource in the graph here? Or make it a config option?

Because, for example in the Backstage backend, if I only have references, I will need to fetch the brokers again. But, we have the full resource here fetched anyway.

Reply:
@Cali0707 this is done, right?

Reply:
Yes, the resources are all added with type `interface{}` to each vertex